tests/unittests/test_exception_handling.py

```diff
@@ -88,158 +60,71 @@ def mocked_internalservererror_500_error(*args, **kwargs):
+        mock_session.side_effect=mock_response
+        error_message = ERROR_CODE_EXCEPTION_MAPPING.get(err_code).get("message")
+        
+        expected_error_message = "HTTP-error-code: {}, Error: {}".format(err_code, error_message)
+        with self.assertRaises(error) as e:
+            client.request(url)
             
-            # Verifying the message formed for the custom exception
-            self.assertEqual(str(e), expected_error_message)
-            pass
-
+        # Verifying the message formed for the custom exception
+        self.assertEqual(str(e.exception), expected_error_message)
 
-    @mock.patch('requests.Request', side_effect=mocked_failed_429_request)
-    def test_too_many_requests_429_backoff_behavior(self, mocked_session, mocked_failed_429_request):
-        config = {'token': '123'}
-        client = client_.Client(config)
-        url = client.build_url(self.endpoint)
-        try:
-            client.request('GET', url)
-        except (requests.HTTPError, client_.TypeformTooManyError) as e:
-            pass
-
-        #Verify daily limit should not backoff
-        self.assertEqual(mocked_failed_429_request.call_count, 3)
-        self.assertEqual(mocked_session.call_count, 3)
+@mock.patch('tap_typeform.client.requests.Session.get')
+class TestClientBackoffHandling(unittest.TestCase):
+    """
+    Test handling of backoff for Timeout, ConnectionError, ChunkEncoding, 5xx, 429 errors.
+    """
 
+    endpoint = "forms"
 
-    @mock.patch('requests.Request', side_effect=mocked_internalservererror_500_error)
-    def test_internalservererror_500_backoff_behaviour(self, mocked_session, mocked_internalservererror_500_error):
+    @parameterized.expand([
+        (requests.exceptions.ConnectionError, requests.exceptions.ConnectionError, 5),
+        (requests.exceptions.Timeout, requests.exceptions.Timeout, 5),
+        (requests.exceptions.ChunkedEncodingError, requests.exceptions.ChunkedEncodingError, 3),
+        (client_.TypeformInternalError, mocked_internalservererror_500_error, 3),
+        (client_.TypeformNotAvailableError, mocked_not_available_503_error, 3),
+        (client_.TypeformTooManyError, mocked_failed_429_request, 3),
+    ])
+    def test_back_off_error_handling(self, mock_session, error,mock_response, expected_call_count):
+        """
+        Test handling of backoff that function is retrying expected times
+        """
+        mock_session.side_effect = mock_response
         config = {'token': '123'}
         client = client_.Client(config)
         url = client.build_url(self.endpoint)
-        try:
-            client.request('GET', url)
-        except (requests.HTTPError, client_.TypeformInternalError) as e:
-            pass
-
-        self.assertEqual(mocked_internalservererror_500_error.call_count, 3)
-        self.assertEqual(mocked_session.call_count, 3)
-
+        with self.assertRaises(error):
+            client.request(url)
```

Comment:
unittests of raw data items Logger is missing for the `request` method.

Reply:
Added unittests